File: src/main/scala/robosky/uplands/client/render/entity/EntityRenderingRegistry.scala
package robosky.uplands.client.render.entity

import net.fabricmc.fabric.api.client.rendereregistry.v1.EntityRendererRegistry
import net.minecraft.client.render.entity.{EntityRenderDispatcher, EntityRenderer}
import net.minecraft.entity.{Entity, EntityType}

import robosky.uplands.entity.EntityRegistry

import scala.reflect.{ClassTag, classTag}

object EntityRenderingRegistry {

  private def register[E <: Entity](tpe: EntityType[E], f: (EntityRenderDispatcher, EntityRendererRegistry.Context) => EntityRenderer[E]): Unit = {
    EntityRendererRegistry.INSTANCE.register(tpe, f(_, _))
  }

  def init(): Unit = {
    register(EntityRegistry.HEXAHAEN, (d, c) => new HexahaenRender(d))
  }
}

File: src/main/java/robosky/uplands/client/render/entity/EntityRenderingRegistry.java
package robosky.uplands.client.render.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.rendereregistry.v1.EntityRendererRegistry;
import net.minecraft.client.render.entity.MobEntityRenderer;
import net.minecraft.entity.EntityType;
import robosky.uplands.entity.EntityRegistry;

@Environment(EnvType.CLIENT)
public class EntityRenderingRegistry {
    public static void init() {
        register(EntityRegistry.HEXAHAEN, HexahaenRender.class);
    }

    private static void register(EntityType<?> entity, Class<? extends MobEntityRenderer<?, ?>> renderer) {
        EntityRendererRegistry.INSTANCE.register(entity, (entityRenderDispatcher, context) -> {
            MobEntityRenderer<?, ?> render = null;
            try {
                render = renderer.getConstructor(entityRenderDispatcher.getClass()).newInstance(entityRenderDispatcher);
            } catch (ReflectiveOperationException e) {
                e.printStackTrace();
            }
            return render;
        });
    }
}
